전남대 FE 최혁 - 3단계 단체 주문 기능 구현하기

aaaaaattt opened the pull request:
## 주문하기 페이지의 폼을 React Hook Form를 사용해서 리팩터링해요. (form 데이터를 useState로 다루는 경우는 없게 해요.)

- [x] React Hook Form (이름,전화번호,수량)

## 받는 사람 기능 고도화

- [x] 참고 URL을 살펴보고 받는 사람 기능의 UI를 구현해요.
- [x] 받는 사람을 최대 10명까지 등록 가능하게 해요.
- [x] 10명의 정보가 정확하게 입력되지 않으면 정보가 반영되지 않아요.
- [x] 받는 사람의 전화번호가 중복되면 안돼요.
- [x] 최소 선택 수량은 1개 이상이어야 해요.
- [x] 전화번호는 01012341234 형태로만 받을 수 있어요.

## 주문하기 페이지 고도화

- [?] 참고 URL을 살펴보고 변경된 기능을 구현해요.
  (선택) 유효성 검사를 Zod 라이브러리를 사용해서 처리해요.

nhdor committed:
refactor:  strict equal 수정

nhdor committed:
refactor: font-size에 theme 적용

nhdor committed:
refactor: '선물하기' 클릭하면 '/' 경로로 이동

nhdor committed:
feat: 로그인 사용자 ProtectedRoute로 검증 추가

nhdor committed:
feat: 과제 진행 요구사항 추가

nhdor committed:
refactor: ProtectedRoute 추가로 인한 같은 기능을 하는 코드들 제거

nhdor committed:
feat: Oreder 컴포넌트의 받는 사람의 입력 폼 React Hook Form으로 구현

nhdor committed:
chore: Order 컴포넌트에서 react hook form의 받는 사람 폼의 구현 부분 제거

nhdor committed:
feat: 폼 제출 완료 버튼 추가

nhdor committed:
feat: 추가 버튼을 눌렀을 때 modal 창 추가와 OrderForm 연동

nhdor committed:
feat: 요구사항 체크 반영

nhdor committed:
feat: 모달창 scroll css 추가

nhdor committed:
chore: 타입 오타 수정

nhdor committed:
feat: 최대 인원 10명을 fileds의 length와 disabled 속성 활용

nhdor committed:
feat: 요구사항 반영
- 10명 인원 제한

nhdor committed:
feat: 타입 검증 에러 수정

nhdor committed:
feat: watch를 활용해서 받는 사람의 전화번호 중복 방지

nhdor committed:
chore: 모달창중복 onClick 삭제

nhdor committed:
feat: ReceiveModal 컴포넌트 분리

nhdor committed:
feat: 받는 사람  테이블 표시 추가

nhdor committed:
feat: table에 css-in-JS 적용

nhdor committed:
feat: 받는 사람의 인원과 주문 금액 연동

nhdor committed:
feat: 주문하기 버튼 클릭시 알림 팝업창 추가

nhdor committed:
feat: 주문 이후 home으로 이동

nhdor committed:
feat: 요구사항체크리스트 반영

youngjewoo commented on src/components/GiftRankingHeader.tsx:
```diff
@@ -52,7 +52,7 @@ const GiftRankingHeader = () => {
             css={[
               iconStyle(theme),
-              target == "ALL" && selectedTargetStyle(theme),
+              target === "ALL" && selectedTargetStyle(theme),
```
👍

youngjewoo commented on src/App.tsx:
```diff
@@ -21,8 +22,23 @@ function App() {
+              element={
+                <ProtectedRoute>
+                  <MyPage />
+                </ProtectedRoute>
```
👍

youngjewoo commented on src/components/NavigationBar.tsx:
```diff
@@ -24,7 +24,14 @@ const NavigationBar = () => {
+        css={titleStyles(theme)}
+      >
+        선물하기
+      </h1>
```
👍

youngjewoo commented on src/components/order/Order.tsx:
```diff
@@ -4,13 +4,12 @@ import { useTheme } from "@emotion/react";
   MessageStyle,
   FormSectionWrapperStyle,
-  ReceiverFormStyle,
+  // ReceiverFormStyle,
```
사용하지 않는 주석은 불태워주셔요 ㅋ

youngjewoo commented on src/components/order/Order.tsx:
```diff
@@ -26,41 +25,52 @@ import {
+import { css } from "@emotion/react";
+import ReceiverModal from "@/components/order/ReceiveModal";
+
+const ReceiverInfoTable: React.FC<{
```
오우 요건 깔끔하게 컴포넌트로 분리돼있네요!

모딱 들어서 별도의 파일로 빼면 좋겠죠? 

하나의 파일에 여러 개 컴포넌트가 있으면 검색으로 찾기가 어렵습니다.

youngjewoo commented on src/components/order/Order.tsx:
```diff
@@ -26,41 +25,52 @@ import {
-  }, []);
+  const [isModalOpen, setIsModalOpen] = useState(false);
+  const [receivers, setReceivers] = useState<
+    { receiverName: string; phoneNumber: string; quantity: number }[]
```
요기는 밑에 선언된 FormData의 타입을 재활용해보면 어떨까요??
타입이 여러 개로 나뉘면 코드 수정 및 데이터 정합성에 불리합니다.

[SSoT(Single Source of Truth)](https://ko.wikipedia.org/wiki/%EB%8B%A8%EC%9D%BC_%EC%A7%84%EC%8B%A4_%EA%B3%B5%EA%B8%89%EC%9B%90) 라고 데이터와 스키마 정의는 한 곳에서 관리해야한다는 권고사항도 있으니
한 번 참고 부탁드려요!

youngjewoo replied:
이 부분도 접혀서 못 읽으신 거 같은데..! 
한 번 확인 부탁드려요!

youngjewoo commented on src/components/order/Order.tsx:
```diff
@@ -26,41 +25,52 @@ import {
   const [selectedId, setSelectedId] = useState<number>();
   const { id } = useParams<{ id: string }>();
-  const [quantity, setQuantity] = useState(1);
+  const [quantity] = useState(1);
```
setter가 없는 state의 선언이 필요할까용??

수정될 필요가 없다면 상수로 선언해도 될 거 같습니다!

youngjewoo commented on package.json:
```diff
@@ -15,6 +15,7 @@
     "pretendard": "^1.3.9",
     "react": "^19.1.0",
     "react-dom": "^19.1.0",
+    "react-hook-form": "^7.60.0",
```
[ change request ] 1. 로그인 상태 오류 (2단계 미션)

<img width="630" height="217" alt="image" src="https://github.com/user-attachments/assets/42711eb6-9e67-420a-b740-7bdcbe313574" />

비밀번호를 8자 이상 입력해서 로그인 버튼이 활성화된 상태임에도
빨간 밑줄이 사라지지 않습니다.

수정 부탁드려요!

[ change request ] 2. 마이페이지에서 새로고침 시 정보 누락 (2단계 미션)

<img width="763" height="213" alt="image" src="https://github.com/user-attachments/assets/eed443c7-3960-4fa8-a5aa-ceeddb495318" />

제가 이전 단계에서 놓쳤던 걸까요..?
sessionStorage에는 정보가 있는데 표시가 되지 않습니다. 한 번 확인해주세요!

youngjewoo commented on src/components/order/OrderForm.tsx:
```diff
@@ -0,0 +1,139 @@
+  });
+
+  const onSubmit = (data: FormData) => {
+    console.log("제출된 데이터:", data);
```
로그는 디버깅용이었다면 제거 부탁드립니다!

youngjewoo commented on src/components/order/OrderForm.tsx:
```diff
@@ -0,0 +1,139 @@
+      <button
+        type="button"
+        onClick={() => {
+          console.log(fields.length);
```
여기도요!

youngjewoo commented on src/components/order/OrderForm.tsx:
```diff
@@ -0,0 +1,139 @@
+
+  const onSubmit = (data: FormData) => {
+    console.log("제출된 데이터:", data);
+    alert("주문 완료!");
```
[ change request ] 3. 요구사항에 어긋나는 alert 제거

<img width="1006" height="860" alt="image" src="https://github.com/user-attachments/assets/27654ad3-592d-49fa-90c8-b129454c803b" />

modal에서 선물 받은 사람을 지정했더니 '주문 완료!'라는 메시지가 노출됩니다.
주문하기 버튼을 눌렀을 때 호출되도록 변경 부탁드려요.

[ change request ] 4. 번호 중복, 10명 이상 추가 가능한 문제

<img width="728" height="595" alt="image" src="https://github.com/user-attachments/assets/b970346f-3203-4aef-a36c-97112510450d" />

Modal을 닫았다가 열었을 때 이미 추가된 사람의 데이터가 전달되지 않아서
중복 확인 및 인원 제한 로직이 제대로 동작하지 않습니다.

변경부탁드려요!

[ change request ] 5. Modal에 Style 적용

강사님 CSS와 전체적으로 너무 다른 부분이 있어서 Modal 내부에도 한 번 스타일을 어느정도 맞춰주시면 감사하겠습니다!

youngjewoo commented on src/components/order/Order.tsx:
```diff
@@ -71,34 +81,21 @@ const Order: React.FC = () => {
+        `보내는 사람: ${SenderNameRef.current?.value}\n` +
+        `메시지: ${GiftMessageRef.current?.value}\n` +
+        `받는 사람 수: ${receivers.length}명\n` +
+        `총 수량: ${receivers.length === 1 ? quantity : quantity * receivers.length}개`
```
[ change request ] 6. 수량과 가격 불일치

<img width="643" height="295" alt="image" src="https://github.com/user-attachments/assets/2ffdf10d-e28d-4ff6-8e99-43ae7c6451d0" />

수량이 100개면 최소 290만원이 넘어야하는데 alert 문과 총액 계산 로직에 문제가 있는 거 같습니다.
수정 부탁드려요!

youngjewoo reviewed (changes requested):
구현하느라 고생하셨습니다!

zod 라이브러리보다는 과제 핵심 기능에 초점을 맞춰서 구현하신 것으로 보이네요!

근데 과제 내용들 중 몇 가지 부분이 누락됐고, 2단계 구현 내용들도 몇 가지 틀어져서
수정 요청 드렸습니다.

확인해보시고 수정해서 다시 올려주셔요!

nhdor committed:
chore:
- 주석 제거
- console.log 제거

nhdor committed:
feat: ReceiverInfoTable 분리

nhdor committed:
fix: Input valid에 따라 input border-line 표시되도록 수정

nhdor committed:
feat: MyPage 새로고침 시 정보 누락 해결
- 원인 : UserInfoProvider에서 user에 session에서 저장된 값이 아닌 null값을 기본으로 설정.
- 해결방법 : 초기 user에 session 저장된 값 할당

nhdor committed:
chore: 컴포넌트 이름 수정

nhdor committed:
feat: 주문폼 style 적용

nhdor committed:
chore: modal 컴포넌트 이름변경

nhdor committed:
style: 보내는 사람 Input css 수정(세로폭)

nhdor committed:
fix: 총 주문 금액 계산 부분 수정
- 변경 전 : 받는 사람의 인원수를 참조해서 계산
- 변경 후 : 주문 수량과 선택된 상품의 판매가로 계산

nhdor committed:
chore: 사용하지 않는 코드 삭제

nhdor committed:
refactor:
- 수신자 정보를 orderForm으로 props 전달
- reset을 활용해서 수신자들이 변경될 때마다 렌더링 되도록

nhdor committed:
chore: 수신자 테이블 재추가

nhdor committed:
chore: style css 수정

nhdor committed:
fix: 수신자 데이터를 새롭게 추가하지 않고 새롭게 다시 set으로 설정하도록 변경

nhdor committed:
fix: 받는 사람의 등록 여부에 따라 버튼의 텍스트 변경

nhdor committed:
feat: Order 페이지에 받는 사람 추가/수정 버튼 스타일 적용

nhdor committed:
fix: 취소화 완료 버튼 같은 component에 위치하도록 수정

nhdor committed:
feat: style css 적용

nhdor committed:
feat: style css적용
- 선물보내기 메시지 부분

aaaaaattt commented:
> 구현하느라 고생하셨습니다!
> 
> zod 라이브러리보다는 과제 핵심 기능에 초점을 맞춰서 구현하신 것으로 보이네요!
> 
> 근데 과제 내용들 중 몇 가지 부분이 누락됐고, 2단계 구현 내용들도 몇 가지 틀어져서 수정 요청 드렸습니다.
> 
> 확인해보시고 수정해서 다시 올려주셔요!

주말까지 소중한 시간 내주시면서 피드백 해주셔서 다시 한번 감사드립니다!
제가 구현하면서 화면에 보이는 부분만 신경썼지 제대로 검증을 하지 않았더니 작성한 부분이 잘못된건지도 모르고 지나갈 뻔 했네요.(제대로 된 검증의 중요성)
또한 여러 미션을 진행하면서 타입을 검증하는 부분과 컴포넌트 간의 상태 공유 또는 데이터의 흐름과 관련된 부분이 제 개인적으로 중요해이고 실제로 제가 어떤 UI를 눌렀을 때 어떤 순서로 진행되면서 흐름에 따라 입력된 값이 어느 컴포넌트로 전달되고 바뀌는지 확인할 수 있는 디버깅 방법이 있을지 궁금해서 질문 남깁니다!

youngjewoo commented on src/components/order/Order.tsx:
```diff
@@ -26,41 +24,33 @@ import {
     const msg = GiftMessageRef.current?.value.trim() ?? "";
     const sender = SenderNameRef.current?.value.trim() ?? "";
-    const receiver = ReceiverNameRef.current?.value.trim() ?? "";
-    const phone = PhoneNumberRef.current?.value.trim() ?? "";
```
[ change request ] 7. 받는 사람이 없어도 선물 발송이 가능한 문제

<img width="715" height="1017" alt="image" src="https://github.com/user-attachments/assets/afafef47-9b6d-49b4-900a-8bc4a17944f2" />

받는 사람을 지정하지 않아도 선물을 발송할 수 있는 문제가 있습니다!
valid 조건에서 한 번 확인 부탁드려용

youngjewoo commented on src/components/order/ReceiverInfoTable.tsx:
```diff
@@ -0,0 +1,44 @@
+import { css } from "@emotion/react";
+
+const ReceiverInfoTable: React.FC<{
```
🥳

youngjewoo commented on src/context/UserInfoProvider.tsx:
```diff
@@ -1,7 +1,7 @@
 
 type User = {
-  email: string | null;
+  email: string;
```
commit 메시지에 버그의 원인과 결과를 적어주는 것이 좋네요!

전 버그 수정 커밋의 경우 이렇게 적어주는 동료를 선호합니다.

youngjewoo commented on src/components/order/Order.tsx:
```diff
@@ -44,6 +44,10 @@ const Order: React.FC = () => {
   const navigate = useNavigate();
+  const totalQuantity =
+    receivers.length === 0
+      ? 1
```
받는 사람이 0이면 수량도 0이어야 하지 않을까.. 싶은데요

일단 과제 구현에 명기된 부분은 아니니 넘어가겠습니다.

aaaaaattt replied:
0이 맞는 것 같습니다! 제가 받는 사람이 없어도 default 값으로 상품 가격이 결제하기 옆에 표시되어야한다고 생각했던 것 같아요!

youngjewoo commented on src/components/MyPage.tsx:
```diff
@@ -1,32 +1,14 @@
 
 const MyPage: React.FC = () => {
-  const { user, setUser } = useUserInfo();
+  const { user } = useUserInfo();
```
[ change request ] 8. useUserInfo hook의 파일 분리

예전에 제가 놓친 거 같은데요..!

저 훅이 `UserInfoProvider.ts` 컨텍스트 정의하는 파일 안에 있습니다.

근데 보통 react에서는 1 file / 1 component 컨벤션을 가져가는 편이라서요!
hook도 일반적으로 그렇습니다.

그래서 IDE에서 warning을 띄워주네요.

<img width="915" height="222" alt="image" src="https://github.com/user-attachments/assets/0aa5d42d-d040-4449-b58d-2408a70b5a05" />

hooks 디렉토리를 만들고 그 쪽으로 코드 이동 부탁드립니다.

[ change request ] 9. data 디렉토리 하위 불필요한 js 파일 제거

이것도 제가 놓쳤던 거 같네요..!

불필요한 js 파일은 제거 부탁드립니다.

youngjewoo commented on src/components/order/ReceiverModal.tsx:
```diff
@@ -0,0 +1,72 @@
+          * 받는 사람의 전화번호를 중복으로 입력할 수 없어요.
+        </p>
+        <OrderForm
+          onSubmitCallback={(formData) => {
```
```suggestion
          onSubmitCallback={onSave}
```

이거랑 같은 거 같은데 결과가 달라지는지 한 번 확인부탁드려용!

youngjewoo commented on src/components/order/ReceiverInputSet.tsx:
```diff
@@ -0,0 +1,134 @@
+          required: "받는 사람의 이름은 필수 입력값입니다.",
+        })}
+      />
+      {Array.isArray(errors.order) && errors.order[index]?.receiverName && (
```
코드 정리 제안

Array.isArray(errors.order) && errors.order[index]?.receiverName

이 비슷한 구문이 반복되는데 Array.isArray 검사하시는 목적이 undefined 여부를 체크하는 걸로 보여서
그 의도가 명확히 드러나게 `Boolean(errors.order)`, `!== undefined`, `!!errors.order` 등으로 체크하면 좋을 거 같고요.

헬퍼 함수로 분리하면 코드 중복을 줄일 수 있을 거 같네요!

[ 예시 ]

```ts
// field의 타입 부분은 별도의 타입으로 분리돼있지 않아서 일단 이렇게 썼는데
// 별도 타입으로 분리해서 그 타입을 넣으시는 걸 추천 드립니다.
const getError = (field: keyof FormData["order"][0]) =>
  Boolean(errors.order?.[index]) ? errors.order[index][field]?.message : null;

// 사용

{ getError('receiverName') && (<p>...</p>) }
```

youngjewoo reviewed (changes requested):
이전 기능이 누락된 부분과 스타일을 잘 보완해주셔서 완성도가 높아진 거 같네요!

두 가지만 더 요청 드리겠습니다.

1. 받는 사람이 지정되지 않아도 주문이 완료 되는 문제
2. hook 파일 분리 -> react fast refresh가 동작하지 않으면 버그처럼 보일 수도 있습니다.

나머지는 필수 사항은 아니니 한 번 읽어주셔요!
이 외 부분들은 나쁘지 않습니다.

그리고  UI를 눌렀을 때 데이터 흐름을 확인할 수 있는 툴이 있냐고 했는데,
제 기억에는 redux 같은 상태 관리 라이브러리를 쓰거나 (flux architecture)
직접 props 전달같은 데이터 흐름을 명시적으로 보이도록 구조를 짜는 것이 중요합니다.

근데 리액트 특성상 useState, useEffect를 사용하다보면 완벽하게 제어하는 것은
예외 케이스가 한 두개씩 발생하기도 하고 그래서 쉽지는 않습니다.

그런 케이스는 제외하더라도 일관적인 규칙들이 있으면 좋겠죠!
그 팁은 아래 두 문서를 참고 해보시면 좋을 거 같습니다.

1. [리액트 state 구조 가이드](https://ko.react.dev/learn/choosing-the-state-structure#avoid-redundant-state)
2. [데이터 중심 애플리케이션 설계](https://www.yes24.com/product/goods/59566585)

2번은 책인데 앞 부분에 참고할만한 내용들이 있습니다!
(일단 지금은 1번만 읽으셔도 무방해요.)

그리고 응원의 한 마디를 적자면 linux 운영체제를 만든 리누스 토발즈가 남긴 명언이 있는데요.

> Good programmers worry about data structures and their relationships
>   --  Linus Torvalds

데이터의 흐름을 고민하고 있는 혁님은 이미 굿 프로그래머라고 할 수 있겠네요.

일단 props top-down 전달 구조 유지, 
prop drilling 이나 부모-자식 모두 state를 수정하고 공유하는 경우 
context 등을 활용해보시면서

내가 state가 얽히면서 겪었던 불편함을 해소할 수 있는지 경험해보시는게 도움이 될 거 같습니다.
(달라지는 건 없지만 시간은 많이들 거에요.. 하지만 그게 고급 개발자로서의 첫 걸음입니다.)

일단 과제 통과가 우선이니 시간이 부족하시면 제가 드린 변경 요청 먼저 반영 해주시고요.
제 설명이 너무 추상적이었다면 질문 또 남겨주세요!

화이팅입니다.

nhdor committed:
refactor: FormData의 타입 재활용으로 receivers의 타입검증(SSOT 고려)

nhdor committed:
fix: 받는 사람이 없어도 선물 발송이 가능한 문제 해결
- receivers의 length를 기준으로 valid 조건 추가

nhdor committed:
fix: 받는 사람이 0이면 수량도 0이 되도록 수정

nhdor committed:
refactor: useUserInfo hook 분리

nhdor committed:
chore: 불필요한 .js 파일 삭제

nhdor committed:
refactor:  불필요한  함수 제거

nhdor committed:
refactor: 코드 정리
- 배열의 undefined 체크  헬퍼 함수로 분리하여 정리

aaaaaattt commented:
> 이전 기능이 누락된 부분과 스타일을 잘 보완해주셔서 완성도가 높아진 거 같네요!
> 
> 
> 
> 두 가지만 더 요청 드리겠습니다.
> 
> 
> 
> 1. 받는 사람이 지정되지 않아도 주문이 완료 되는 문제
> 
> 2. hook 파일 분리 -> react fast refresh가 동작하지 않으면 버그처럼 보일 수도 있습니다.
> 
> 
> 
> 나머지는 필수 사항은 아니니 한 번 읽어주셔요!
> 
> 이 외 부분들은 나쁘지 않습니다.
> 
> 
> 
> 그리고  UI를 눌렀을 때 데이터 흐름을 확인할 수 있는 툴이 있냐고 했는데,
> 
> 제 기억에는 redux 같은 상태 관리 라이브러리를 쓰거나 (flux architecture)
> 
> 직접 props 전달같은 데이터 흐름을 명시적으로 보이도록 구조를 짜는 것이 중요합니다.
> 
> 
> 
> 근데 리액트 특성상 useState, useEffect를 사용하다보면 완벽하게 제어하는 것은
> 
> 예외 케이스가 한 두개씩 발생하기도 하고 그래서 쉽지는 않습니다.
> 
> 
> 
> 그런 케이스는 제외하더라도 일관적인 규칙들이 있으면 좋겠죠!
> 
> 그 팁은 아래 두 문서를 참고 해보시면 좋을 거 같습니다.
> 
> 
> 
> 1. [리액트 state 구조 가이드](https://ko.react.dev/learn/choosing-the-state-structure#avoid-redundant-state)
> 
> 2. [데이터 중심 애플리케이션 설계](https://www.yes24.com/product/goods/59566585)
> 
> 
> 
> 2번은 책인데 앞 부분에 참고할만한 내용들이 있습니다!
> 
> (일단 지금은 1번만 읽으셔도 무방해요.)
> 
> 
> 
> 그리고 응원의 한 마디를 적자면 linux 운영체제를 만든 리누스 토발즈가 남긴 명언이 있는데요.
> 
> 
> 
> > Good programmers worry about data structures and their relationships
> 
> >   --  Linus Torvalds
> 
> 
> 
> 데이터의 흐름을 고민하고 있는 혁님은 이미 굿 프로그래머라고 할 수 있겠네요.
> 
> 
> 
> 일단 props top-down 전달 구조 유지, 
> 
> prop drilling 이나 부모-자식 모두 state를 수정하고 공유하는 경우 
> 
> context 등을 활용해보시면서
> 
> 
> 
> 내가 state가 얽히면서 겪었던 불편함을 해소할 수 있는지 경험해보시는게 도움이 될 거 같습니다.
> 
> (달라지는 건 없지만 시간은 많이들 거에요.. 하지만 그게 고급 개발자로서의 첫 걸음입니다.)
> 
> 
> 
> 일단 과제 통과가 우선이니 시간이 부족하시면 제가 드린 변경 요청 먼저 반영 해주시고요.
> 
> 제 설명이 너무 추상적이었다면 질문 또 남겨주세요!
> 
> 
> 
> 화이팅입니다.

어떻게 보면 구조를 잘 설계하면 데이터 흐름을 더 명시적으로 확인하면서 진행할 수 있겠군요! 참고로 적어주신 레퍼런스와 '데이텨 중심 애플리케이션 설계' 도서도 한번 읽어봐야겠네요.
말씀해주신 내용 들으니 경험도 중요힌 것 같다는 생각이 드네요 이론적으로만 문제라고 하는 것보다 직접 부딪히면서 개선사항이랑 문제를 찾다보면 설명으로 들었던 내용들이 이미 구현되어있어서 그럴 때 더 어떤 개념이 잘 이해가 되는 것 같았습니다.
이번주에 피드백만 해주셔도 정말 바쁘셨을텐데 제가 고민했던 지점들도 피드백을 주셔서 정말 감사합니다!

youngjewoo commented on src/components/order/Order.tsx:
```diff
@@ -39,9 +39,7 @@ const Order: React.FC = () => {
-  const [receivers, setReceivers] = useState<
-    { receiverName: string; phoneNumber: string; quantity: number }[]
-  >([]);
+  const [receivers, setReceivers] = useState<FormData["order"]>([]);
```
해보니까 타이핑 수도 훨씬 적고, 타입 변경에 훨씬 유연하게 대응할 수 있겠죠..?

이런 경험을 통한 긍정적인 효과를 경험해야 다음에 다시 또 찾게되는 습관이 형성되는 거라고 생각해요.
제 개인적인 경험으로는 책에서 하라니까 해봐야지.. 정도로는 습관까지 만들어지는 거는 힘들더라고요. 또 실제로 그 정도의 유익함을 주지 않으면 습관으로 만들 필요도 없다고 생각합니다..!

이제 보기 흡족하네요

<img width="212" height="166" alt="image" src="https://github.com/user-attachments/assets/a4a6de16-bdbe-4fa9-8762-741d7db9eae0" />

youngjewoo commented on src/components/order/Order.tsx:
```diff
@@ -65,6 +67,16 @@ const Order: React.FC = () => {
+      setReceiverError(
+        "받는 사람이 등록되지 않았습니다. 최소 1명을 선택해주세요."
+      );
+      isValid = false;
```
setMessageErorr, setReceiverError 등등을 저장하니까 isValid 정보는 변수로 저장하지 않고 동적으로 계산할 수도 있을 거 같은데요!
나중에 한 번 도전해보셔요!

제가 저번에 말씀 드린 React state 가이드에 [불필요한 state 피하기](https://ko.react.dev/learn/choosing-the-state-structure#avoid-redundant-state)에도 있는 내용입니다.

youngjewoo commented on src/hooks/useUserInfo.tsx:
```diff
@@ -0,0 +1,10 @@
+import { useContext } from "react";
```
👍

youngjewoo commented on src/components/order/ReceiverInputSet.tsx:
```diff
@@ -69,8 +72,8 @@ const ReceiverInputSet = ({
-      {Array.isArray(errors.order) && errors.order[index]?.phoneNumber && (
-        <p css={errorStyle}>{errors.order[index].phoneNumber?.message}</p>
+      {getError("phoneNumber") && (
+        <p css={errorStyle}>{getError("phoneNumber")}</p>
```
허허이.. 이 사람 밑에 quantity는 빼먹다니오..

aaaaaattt replied:
하하하...

youngjewoo commented on src/context/UserInfoProvider.tsx:
```diff
@@ -1,32 +1,26 @@
 };
 
-const UserInfoContext = createContext<UserInfoContextType | undefined>(
+export const UserInfoContext = createContext<UserInfoContextType | undefined>(
```
[ change request ] 10. context 별도 파일 분리

<img width="913" height="168" alt="image" src="https://github.com/user-attachments/assets/615e43c9-1aa0-4c27-ba25-fae866f12d5e" />

동일하게 fast refresh 이슈가 있습니다!

youngjewoo commented on src/components/login/Login.tsx:
```diff
@@ -31,15 +31,15 @@ const Login = () => {
           onChange={email.onChange}
           onBlur={email.onBlur}
-          css={inputStyle(theme)}
+          css={inputStyle(theme, email.isValid)}
```
[ change request ] 11. useUserInfo 경로 수정 (빌드 에러)

에고 훅으로 분리는 잘 하셨는데.. 경로 다시 수정 부탁드려요...!

이제 이번 미션이 지나가면 다음 멘토분께 넘어가는데 빌드 에러가 있는 상태로 넘길 수는 없습니다 .. ㅠ

 그리고 현업에서는 코드 병합 전 빌드, 테스트, lint 등등을 CI/CD 서버에서 검사하고, 통과한 경우에만 코드 병합이 가능합니다. (잘 관리되는 부서라면요) 구조적 이상이 있는지 탐지는 그 이후의 단계고, 빌드/테스트 오류는 코드 안정성을 크게 떨어뜨리는 행위이기 때문에 PR 병합의 최소조건이라고 할 수 있어요!

 그래서 지금부터 PR을 올리기전 build, lint 등을 한 번 수행해보고 PR을 올리시는 습관을 들여보면 좋을 거 같습니다. 아까 데이터 흐름과 관련해서 경험이 중요한 거 같다고 한 번 언급 주셨는데 소프트웨어 개발에서 일맥상통하는 아주 중요한 포인트입니다. 코드 품질을 유지하기 위해 수동으로 빌드를 돌리며 고통스러워본 경험이 있다면 "에이 이렇게 맨날 할 바에는 조금 고생해서  CI/CD 서버로 자동화 해야지 ㅡㅡ" 라는 사고로 자연스럽게 이어질 수가 있어요. 그러니까 조금 돌아가는 것 같아 보여도 빌드, lint 에러는 한 번 확인하고 올려주세요! 과제 달성에는 돌아가는 길처럼 보이지만 동료에게 인정받는 개발자로 가기에는 지름길입니다.

멘토님들 중에도 빌드 확인을 따로 안 해주시는 분들이 있는 거 같아서 ㅠ 학습에 도움 되시라고 의견 남깁니다!

aaaaaattt replied:
앞으로 npm run lint && npm run build 로 제출 전체 확인하고 요청 드려야겠네요. 지금까지는 중간 과정에서만 확인하고,  눈으로 확인하고 제출하고 해서 에러나 경고가 나오지 않으면 괜찮다고 생각했는데 실제로 실행시키니까 제대로 나오는 걸 이제야 확인했네요ㅠ 좋은 습관으로 한번 만들어보겠습니다! (데이터 흐름보다 어떻게 보면 더 기본적인 내용인 것 같은데 제가 신경을 다른 곳에 쓰고 있었던 느낌이...)
감사합니다~!

youngjewoo replied:
아니에요 ㅋㅋ 다 중요합니다.. 단시간에 모두 익히는 사람은 없고요.
하나하나 배워가는 거죠..

축구에서 패스 드리블 슛 트래핑 중에 어느하나 중요하지 않은 것이 있던가요.. ⚽️

youngjewoo reviewed (changes requested):
에고 확인해봣는데 빌드,린트 에러가 있어서 change request 드렸습니다!

중요한 포인트라서 코멘트 함께 남겼구요! 결과적으로 어려운 수정은 아니실 거에요. 
완료 되시면 바로 리뷰 요청 주셔요!

nhdor committed:
fix: build test 에러
- import 경로 수정

aaaaaattt commented:
> 에고 확인해봣는데 빌드,린트 에러가 있어서 change request 드렸습니다!
> 
> 중요한 포인트라서 코멘트 함께 남겼구요! 결과적으로 어려운 수정은 아니실 거에요. 완료 되시면 바로 리뷰 요청 주셔요!

넵! 바로 반영해보겠습니다

nhdor committed:
refactor: 불필요한 상태 관리 제거
- 변경 전 : 새로운 isValid 변수를 통해 error 검증
- 변경 후 :
이미 존재하는  state를 활용하여 Error 검증

nhdor committed:
chore: 수량 에러 메시지 부분에도 헬퍼함수 적용

nhdor committed:
refactor: userInfoContext 분리
-  fast refresh 이슈 해결

nhdor committed:
chore: imoprt 경로 수정

nhdor committed:
refactor: ESlint test warning 해결
- useEffect 내부에 외부값 의존성 배열에 추가

youngjewoo commented on src/components/GiftRankingHeader.tsx:
```diff
@@ -24,7 +24,7 @@ const GiftRankingHeader = () => {
     params.set("rankType", initRank);
     navigate(`${location.pathname}?${params.toString()}`);
-  }, []);
+  }, [location.pathname, location.search, navigate]);
```
👍

youngjewoo reviewed (approved):
축하드립니다. 이제 드디어 병합할 수 있는 상태가 됐네요!

다른 분보다 여러 번 수정과 검토가 오갔는데요. 그만큼 다른 사람이 봤을 때 좋은 코드가 어떨지에 대해 조금 더 배워가시는 계기가 되셨기를 바라요. (다른 사람에는 1개월 후의 나도 포함됩니다.)

저는 이번 미션까지지만, 다음 미션에도 지금과 같은 열정 보여주세요!
긴 시간동안 고생 많으셨습니다.

aaaaaattt commented:
> 축하드립니다. 이제 드디어 병합할 수 있는 상태가 됐네요!
> 
> 
> 
> 다른 분보다 여러 번 수정과 검토가 오갔는데요. 그만큼 다른 사람이 봤을 때 좋은 코드가 어떨지에 대해 조금 더 배워가시는 계기가 되셨기를 바라요. (다른 사람에는 1개월 후의 나도 포함됩니다.)
> 
> 
> 
> 저는 이번 미션까지지만, 다음 미션에도 지금과 같은 열정 보여주세요!
> 
> 긴 시간동안 고생 많으셨습니다. 

하하 제가 당연히 확인해야할 사항도 놓쳐버리고 안좋은 습관이 있었던 것 같은데 빌드랑 lint 테스트 뿐만 아니라 타입과 관련된 고민, 좋은 코드에 대해서 어렴풋이 생각했던게 이전보다 더 선명해진 3주차였던 것 같습니다. 늦은 시간에도 피드백을 레퍼런스를 바탕으로 자세히 남겨주셔서 여러번 읽으면서 앞으로 방향성을 잡는데 잘 활용해볼게요! 
중요한 시간 내주신만큼 앞으로 제대로 된 방향으로 여러 경험으로 성장하는 학습자가 되겠습니다!
다시 한번 감사드립니다~
